common/flusher/flusher.go

```diff
@@ -0,0 +1,244 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//Type Generic Flush Buffer that is size bound and time bound.
+//The Flush Buffer will flush after a configurable amount of time as well as once the buffer reaches a configurable capacity.
+//The number of flush buffers can also be configured.
+//Starts with x free buffers, once a free buffer reaches capacity or if the timer is up, the free buffer will get switched to a full buffer.
+//A full buffer will get flushed in the background and switched back to a free buffer.
+//When a free buffer switches to a full buffer, another free buffer will take its place if there are any available at the moment.
+
+package flusher
+
+import (
+	"fmt"
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/server/common/channel"
+	"go.temporal.io/server/common/future"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"sync"
+	"time"
+)
+
+type (
+	Flusher[T any] struct {
+		flushDuration     int
+		bufferCount       int
+		bufferCapacity    int
+		flushNotifierChan chan struct{}
+		logger            log.Logger
+		shutdownChan      channel.ShutdownOnce
+		writer            *itemWriterImpl[T]
+
+		sync.Mutex
+		flushTimer       *time.Timer
+		flushItemsBuffer []FlushItem[T]
+		fullBufferChan   chan []FlushItem[T]
+		freeBufferChan   chan []FlushItem[T]
+	}
+	FlushItem[T any] struct {
+		item T
+		fut  *future.FutureImpl[struct{}]
+	}
+	ItemWriter[T any] interface {
+		Write(item ...T) error
+	}
+	itemWriterImpl[T any] struct {
+		logger log.Logger
+		sync.Mutex
+	}
+)
+
+const (
+	bufferCapacityGapPercent = 0.9
+)
+
+func NewFlusher[T any](
+	bufferCount int,
+	bufferCapacity int,
+	flushDuration int,
+	writer *itemWriterImpl[T],
+	logger log.Logger,
+	shutdownChan channel.ShutdownOnce,
+) *Flusher[T] {
+	return &Flusher[T]{
+		flushDuration:     flushDuration,  // time waited after first item insertion before flushing the buffer
+		bufferCount:       bufferCount,    // no of total flush buffers
+		bufferCapacity:    bufferCapacity, // buffer size, will flush a buffer once no of items added to the buffer nears this limit
+		flushTimer:        nil,
+		flushNotifierChan: make(chan struct{}, 1),
+		writer:            writer,
+		flushItemsBuffer:  make([]FlushItem[T], 0, bufferCapacity),
+		fullBufferChan:    make(chan []FlushItem[T], bufferCount),
+		freeBufferChan:    make(chan []FlushItem[T], bufferCount),
+		logger:            logger,
+		shutdownChan:      shutdownChan,
+	}
+}
+
+func NewItemWriterImpl[T any](logger log.Logger) *itemWriterImpl[T] {
+	return &itemWriterImpl[T]{
+		logger: logger,
+	}
+}
+
+func (iwi *itemWriterImpl[T]) Write(flushItem ...FlushItem[T]) error {
+	iwi.Lock()
+	defer iwi.Unlock()
+	return nil
+}
+
+func (f *Flusher[T]) Start() error {
```

Comment:
why return error

Reply:
Was following an example of Start Stop method in another file. I've changed it to return nothing.